"missing outer accessor" if the source file name is alphabetically is before O1.scala

This was originally reported by @xuwei-k as https://github.com/sbt/sbt/issues/7777, using 3.3.4. I've reproduced here without depending on sbt.

## Compiler version

Scala 3.8.0-RC5

## Minimized code

Two source files.

#### A1.scala

```scala
package example

trait A1 { self: O1.type =>
  import O2.someKey
  def taskImpl: Unit = println(someKey)
}
```

#### O1.scala

```scala
package example

object O1 extends A1:
  object O2:
    val someKey = Int
end O1
```

## Output (click arrow to expand)

<details>
<summary>
java.lang.AssertionError: assertion failed: missing outer accessor in package <root>
  | => sat scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
  | => sat dotty.tools.dotc.transform.ExplicitOuter$.dotty$tools$dotc$transform$ExplicitOuter$$$outerParamAccessor(ExplicitOuter.scala:236)
</summary>



```scala
sbt:sbt-2-missing-outer-accessor> compile
[info] compiling 2 Scala sources to /private/tmp/sbt-2-missing-outer-accessor/target/scala-3.8.0-RC5/classes ...
java.lang.AssertionError: assertion failed: missing outer accessor in package <root>
  | => sat scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
  | => sat dotty.tools.dotc.transform.ExplicitOuter$.dotty$tools$dotc$transform$ExplicitOuter$$$outerParamAccessor(ExplicitOuter.scala:236)
	at dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.loop$1(ExplicitOuter.scala:460)
	at dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.path$extension(ExplicitOuter.scala:469)
	at dotty.tools.dotc.ast.tpd$.followOuterLinks(tpd.scala:512)
	at dotty.tools.dotc.ast.tpd$.ref(tpd.scala:484)
	at dotty.tools.dotc.ast.tpd$.desugarIdentPrefix(tpd.scala:1515)
	at dotty.tools.dotc.ast.tpd$.desugarIdent(tpd.scala:1503)
	at dotty.tools.backend.jvm.DottyBackendInterface.cachedDesugarIdent(DottyBackendInterface.scala:34)
	at dotty.tools.backend.jvm.DottyBackendInterface$DesugaredSelect$.unapply(DottyBackendInterface.scala:59)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:762)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:385)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:303)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.loop$1(BCodeBodyBuilder.scala:1209)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments(BCodeBodyBuilder.scala:1216)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:835)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:385)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:899)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:922)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:697)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen$$anonfun$1(BCodeSkelBuilder.scala:705)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:327)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:705)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:293)
	at dotty.tools.backend.jvm.CodeGen.genClass(CodeGen.scala:160)
	at dotty.tools.backend.jvm.CodeGen.genClassDef$1(CodeGen.scala:61)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:117)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1$$anonfun$1(CodeGen.scala:115)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:327)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:115)
	at dotty.tools.backend.jvm.CodeGen.genUnit(CodeGen.scala:120)
	at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:90)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:386)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:327)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:379)
	at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:98)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:380)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1327)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:373)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:420)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:420)
	at scala.Function0.apply$mcV$sp(Function0.scala:45)
	at dotty.tools.dotc.Run.showProgress(Run.scala:482)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:420)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:432)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:432)
	at dotty.tools.dotc.Run.compileSources(Run.scala:319)
	at dotty.tools.dotc.Run.compile(Run.scala:304)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:196)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:252)
```
</details>

## Note

If we rename the source file from `A1.scala` to `O2.scala` or `P1.scala` without changing the content, the compilation succeeds.

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -1520,7 +1520,19 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
         case mt: MethodType if mt.paramInfos.isEmpty && mt.resultType.typeSymbol.is(Module) =>
           ref(mt.resultType.typeSymbol.sourceModule)
         case _ =>
-          ref(prefix)
+          val psym = prefix.symbol
+          if ctx.erasedTypes && psym.is(Module) && psym.isStatic then
+            // After erasure (i.e. when generating code), reference a statically
+            // reachable module through its own (static) path rather than through
+            // `prefix`, whose own prefix may be a `this` that is not addressable here --
+            // e.g. when the module is reached via the self type of a trait, `prefix` is
+            // `Trait.this.Module`, and building an outer path to `Trait.this` fails.
+            // The simplification is only sound for code generation, so it is guarded by
+            // `ctx.erasedTypes` to leave earlier purity/stability analysis untouched.
+            // See i24936.
+            ref(psym)
+          else
+            ref(prefix)
     case TermRef(prefix: ThisType, _) =>
       This(prefix.cls)
     case _ =>

--- a/tests/pos/i24936/A1.scala
+++ b/tests/pos/i24936/A1.scala
@@ -1,0 +1,7 @@
+// https://github.com/scala/scala3/issues/24936
+package example
+
+trait A1 { self: O1.type =>
+  import O2.someKey
+  def taskImpl: Unit = println(someKey)
+}

--- a/tests/pos/i24936/O1.scala
+++ b/tests/pos/i24936/O1.scala
@@ -1,0 +1,6 @@
+package example
+
+object O1 extends A1:
+  object O2:
+    val someKey = Int
+end O1

